File: src/Python/anywave/network/tcprequest.py

from PyQt5 import QtCore, QtNetwork
import anywave
import json
"""
"""
# constants
STATUS_SUCCESS = 0
STATUS_FAILED  = 1
SIZE_INT = 4

class TCPRequest:
    WAIT_TIME_OUT = 3000000 # 300s socket time out
    def __init__(self, request):
        self.request = request
        self.socket = QtNetwork.QTcpSocket()
        self.size = QtCore.QByteArray()
        self.data = QtCore.QByteArray()
        self.streamData = QtCore.QDataStream(self.data, QtCore.QIODevice.WriteOnly)
        self.streamData.setVersion(QtCore.QDataStream.Version.Qt_4_4)
        self.streamSize = QtCore.QDataStream(self.size, QtCore.QIODevice.WriteOnly)
        self.streamSize.setVersion(QtCore.QDataStream.Version.Qt_4_4)
        self.streamResponse = QtCore.QDataStream(self.socket)
        self.streamResponse.setVersion(QtCore.QDataStream.Version.Qt_4_4)
        self.connect()

    def response(self):
        stream = QtCore.QDataStream(self.socket)
        stream.setVersion(QtCore.QDataStream.Qt_4_4)
        return stream

    def connect(self):
        self.socket.connectToHost(anywave.host, anywave.port)
        if not self.socket.waitForConnected():
           self.status = STATUS_FAILED
           error = "Unable to connect to AnyWave: " + self.socket.errorString()
           raise Exception(error)
        self.status = STATUS_SUCCESS

    def waitForResponse(self):
        while self.socket.bytesAvailable() < 4:
            if not self.socket.waitForReadyRead(self.WAIT_TIME_OUT):
                return -1
        stream = QtCore.QDataStream(self.socket)
        stream.setVersion(QtCore.QDataStream.Qt_4_4)
        status = stream.readInt32()
        # get size of data
        while self.socket.bytesAvailable() < 4:
            if not self.socket.waitForReadyRead(self.WAIT_TIME_OUT):
                raise Exception('Nothing to read from socket.')
        size = stream.readInt32()
        # wait for all data to be available
        while self.socket.bytesAvailable() < size:
             if not self.socket.waitForReadyRead(self.WAIT_TIME_OUT):
                    size = -1
        if status is -1:
            error = stream.readQString()
            raise Exception(error)
        if size is -1:
            raise Exception('Data waiting timed out.')
        return size

    def simpleRequest(self):
        # arg is a string that could be empty
        if not self.status == STATUS_SUCCESS:
            raise Exception('not connected to AnyWave')
            return False
        self.streamSize << anywave.pid << int(SIZE_INT) << self.request
        self.socket.write(self.size)
        self.socket.flush()
        if not self.socket.waitForBytesWritten():
            raise Exception('Error while sending request to AnyWave')
            return False
        return True

    # sendRequest() send a request with parameters a json string.
    # the string can be none or empty
    def sendRequest(self, args):
        # args must be a dict
        if self.status is not STATUS_SUCCESS:
            raise Exception('Sending request while not connected to AnyWave')
            return False
        str = json_dumps(args)
        self.streamSize << anywave.pid << str.size() + SIZE_INT << self.request
        self.socket.write(self.size)
        self.streamData << json
        self.socket.write(self.data)
        self.socket.flush()
        if not self.socket.waitForBytesWritten():
            raise Exception('Error while sending request to AnyWave')
            return False
        return True

    # sendString() will only send a string to AnyWave
    # the string cand be a text message or a json string depending on the command
    def sendString(self, str):
        if self.status is not STATUS_SUCCESS:
           raise Exception('Sending request while not connected to AnyWave')
           return False
        self.streamSize.writeInt32(anywave.pid)
        self.streamSize.writeInt32(int(len(str) + SIZE_INT))
        self.streamSize.writeInt32(self.request)
        self.streamData.writeQString(str)
        self.socket.write(self.size)
        self.socket.write(self.data)
        if not self.socket.waitForBytesWritten():
            raise Exception('Error while sending request to AnyWave')
            return False
        return True
    # sendData() will send binary data to anywave.
    # data must be bytes
    def sendData(self, data):
        # data must be a bytes, not a json string
        if self.status is not STATUS_SUCCESS:
            raise Exception('Sending request while not connected to AnyWave')
            return False
        if not isinstance(data, bytes):
            raise Exception('argument must be of type bytes')
            return False
        # use a qbyte array to send the data
        self.streamSize.writeInt32(anywave.pid)
        self.streamSize.writeInt32(int(len(data) + SIZE_INT))
        self.streamSize.writeInt32(self.request)
        self.streamData.writeBytes(data)
        self.socket.write(self.size)
        self.socket.write(self.data)
        #self.socket.flush()
        if not self.socket.waitForBytesWritten():
            raise Exception('Error while sending request to AnyWave')
            return False
        return True

File: src/matlab/MATLAB_Interface/TCPRequest.cpp
#include "TCPRequest.h"
#include "MATLAB_Interface.h"

TCPRequest::TCPRequest(int request)
{
	m_request = request;
	m_status = TCPRequest::Success;
	m_streamData = new QDataStream(&m_data, QIODevice::WriteOnly);
	m_streamSize = new QDataStream(&m_size, QIODevice::WriteOnly);
	m_streamSize->setVersion(QDataStream::Qt_4_4);
	m_streamData->setVersion(QDataStream::Qt_4_4);
	m_streamResponse = new QDataStream(&m_socket);
	m_streamResponse->setVersion(QDataStream::Qt_4_4);
	connect();
}

TCPRequest::~TCPRequest()
{
	delete m_streamData;
	delete m_streamSize;
	delete m_streamResponse;
}

void TCPRequest::connect()
{
	m_socket.connectToHost(MATLAB_Interface::host, MATLAB_Interface::port);
	if (!m_socket.waitForConnected()) {
		m_status = TCPRequest::Failed;
		m_errorString = QString("Failed to connect to AnyWave: %1\n").arg(m_socket.errorString());
		m_errorString += QString("Host: %1 Port: %2").arg(MATLAB_Interface::host).arg(MATLAB_Interface::port);
		throw(m_errorString);
	}
	else
		m_status = TCPRequest::Success;
}

void TCPRequest::clear()
{
	m_data.clear();
	m_size.clear();
	m_streamData->device()->reset();
	m_streamSize->device()->reset();
}

int TCPRequest::waitForResponse()
{
	while (m_socket.bytesAvailable() < 4)
		if (!m_socket.waitForReadyRead(WAIT_TIME_OUT))
			return -1;
	QDataStream* response = new QDataStream(&m_socket);
	response->setVersion(QDataStream::Qt_4_4);
	int status;
	*response >> status;
	if (status == -1) {
		*response >> m_errorString;
		throw(m_errorString);
	}
	while (m_socket.bytesAvailable() < 4)
		if (!m_socket.waitForReadyRead(WAIT_TIME_OUT)) {
			m_errorString = "Nothing to read from socket.";
			throw(m_errorString);
		}
	int size;
	*response >> size;
	while (m_socket.bytesAvailable() < size)
		if (!m_socket.waitForReadyRead(WAIT_TIME_OUT)) 
			size = -1;
	if (size == -1) {
		m_errorString = "No data received.";
		throw(m_errorString);
	}	
	return size;
}

void TCPRequest::simpleRequest()
{
	if (m_status != TCPRequest::Success) {
		m_errorString = "Not connected to AnyWave.";
		throw(m_errorString);
	}
	clear();
	*m_streamSize << MATLAB_Interface::pid <<  int(sizeof(int)) << m_request;
	m_socket.write(m_size);
	if (!m_socket.waitForBytesWritten()) {
		m_errorString = "Error while sending data to AnyWave.";
		throw(m_errorString);
	}

}

void TCPRequest::sendString(const QString& s)
{
	clear();
	*m_streamData << s;
	// write the total size of data in size: total size is in fact  data size + size of request value (int)
	// first 4 bytes of size QBytesArray MUST BE the pid number of process.
	*m_streamSize << MATLAB_Interface::pid << int(m_data.size() + sizeof(int)) << m_request;
	m_socket.write(m_size);
	m_socket.write(m_data);
	if (!m_socket.waitForBytesWritten()) {
		m_errorString = "Error while sending data to AnyWave.";
		throw(m_errorString);
	}
}

void  TCPRequest::sendData(const QByteArray& data)
{
	clear();
	// write the total size of data in size: total size is in fact  data size + size of request value (int)
	// first 4 bytes of size QBytesArray MUST BE the pid number of process.
	*m_streamSize << MATLAB_Interface::pid << int(data.size() + sizeof(int)) << m_request;
	*m_streamData << data;
	m_socket.write(m_size);
	m_socket.write(m_data);
	
	if (!m_socket.waitForBytesWritten()) {
		m_errorString = "Error while sending data to AnyWave.";
		throw(m_errorString);
	}
}
